### PR TITLE
clh: Support VFIO device unplug

### PR DESCRIPTION
**Pull request by likebreath**
This patch adds the support of VFIO device unplug when using
cloud-hypervisor.

Fixes: #2569

Signed-off-by: Bo Chen <chen.bo@intel.com>

### CODE REVIEW

File: virtcontainers/pkg/cloud-hypervisor/Makefile [hunk 1]
```diff
@@ -13,7 +13,7 @@ YQ := $(shell command -v yq 2> /dev/null)
 generate-client-code: clean-generated-code
 	docker run --rm \
 		--user $$(id -u):$$(id -g) \
-		-v $${PWD}:/local openapitools/openapi-generator-cli generate \
+		-v $${PWD}:/local openapitools/openapi-generator-cli:v4.3.1 generate \
```
Comment (likebreath) [inline virtcontainers/pkg/cloud-hypervisor/Makefile:16]: /cc @devimc I added the same tag to the `openapi-generator-cli` container as you did in the SGX PR in kata 2.0 (https://github.com/kata-containers/kata-containers/pull/484). No separate back-porting PR is needed for tagging the container. :)

**Comment by likebreath** [comment]
As cloud-hypervisor provides a united way to remove devices (including `blk/pmem/net/vfio`, etc), I will also add another commit to simply our `hotplugRemove*` functions.

File: virtcontainers/pkg/cloud-hypervisor/Makefile (hunk 1 above)
Comment (devimc) [reply virtcontainers/pkg/cloud-hypervisor/Makefile:16]: thanks @likebreath - I appreciate it :+1:

**Review by devimc — APPROVED** [review]
VFIO CI is happy - I'm happy too :smile: 
thanks @likebreath

**Review by jodh-intel — APPROVED** [review]
Thanks @likebreath.

lgtm

**Comment by jodh-intel** [comment]
@likebreath - I'm assuming this is a feature, or is it actually a bug fix? Could you decide and then apply the appropriate backport label to show whether we need to port to the stable branches. We can then land this :)

**Comment by likebreath** [comment]
@jodh-intel Thanks for the reminder. Yes, it is a feature, so I think it won't be needed for back-porting.

**Comment by likebreath** [comment]
The only relevant CI failing is for the `clh-metrics`. It is failing on boot time, as we are booting faster than expected from the reference metrics data ([here](https://github.com/kata-containers/tests/blob/master/cmd/checkmetrics/ci_slaves/checkmetrics-json-clh-baremetal-kata-metric3.toml#L19)). The reduced boot time is expected from the PR, as we removed couple of unecessary `VMMPing`. Quotes from the failing CI log:
Job 128: `02:08:00 time="2020-10-01T09:08:00Z" level=warning msg="Failed Minval ( 1.200000(minimal mean expected) > 1.079900(minimal mean on dataset)) for [boot-times]"`
Job 129: `07:06:29 time="2020-10-01T14:06:28Z" level=warning msg="Failed Minval ( 1.200000(minimal mean expected) > 1.045350(minimal mean on dataset)) for [boot-times]"`
Job 130: `10:45:46 time="2020-10-01T17:45:46Z" level=warning msg="Failed Minval ( 1.200000(minimal mean expected) > 1.058000(minimal mean on dataset)) for [boot-times]"`

I am assuming the way to fix the metrics-ci is to update the reference metrics data ([here](https://github.com/kata-containers/tests/blob/master/cmd/checkmetrics/ci_slaves/checkmetrics-json-clh-baremetal-kata-metric3.toml#L19)). What's the current practice to collect the new metrics data? Will the above 3 executions of the metrics CI be enough to update the data? @GabyCT @chavafg @jcvenegas @amshinde 


Also, the `0.35%` code coverage decrease should be attributed to the new function added, which will be refactored as I mentioned earlier (probably in a separate PR).

**Comment by GabyCT** [comment]
@likebreath yes you can update the metrics limits and then run a `Depends-on` on this PR to see that is working...most of the time I ran them at least > 5 times to get a proper number and then I ran multiple times the metrics CI to see that we have a stable result

**Comment by likebreath** [comment]
> most of the time I ran them at least > 5 times to get a proper number and then I ran multiple times the metrics CI to see that we have a stable result

Thanks @GabyCT. The PR is up here: https://github.com/kata-containers/tests/pull/2912.  I will do the same as you did, and ping you when it is ready for reviewing/merging.

**Comment by GabyCT** [comment]
@jcvenegas and @likebreath I have seen that the podman clh CI is failing with this error
```
02:46:42 HEAD is now at 6d30fe05 vmm: openapi: Add the 'iommu' and 'id' option to 'VmAddDevice'
02:46:42 ./scripts/dev_cli.sh: line 122: docker: command not found
```
It seems that is looking for docker but if I remembered correctly the image does not have docker supported so maybe that is something that needs to be fixed

**Comment by likebreath** [comment]
> ```
> 02:46:42 HEAD is now at 6d30fe05 vmm: openapi: Add the 'iommu' and 'id' option to 'VmAddDevice'
> 02:46:42 ./scripts/dev_cli.sh: line 122: docker: command not found
> ```
> 
> It seems that is looking for docker but if I remembered correctly the image does not have docker supported so maybe that is something that needs to be fixed

@GabyCT Yes, the CI has been broken because of missing docker in the guest image for a while. Do you have any insights on how it can be fixed? Thank you.

**Comment by likebreath** [comment]
Merging the PR, as the clh metrics now is passing.

**Comment by GabyCT** [comment]
@likebreath I believe that fedora by default has now podman but if you want to enable docker you need to modify a flag if I remembered correctly in the kernel (I do not remembered correctly) ..I am not quite sure how to handle that in the CI as the current image has the kernel configuration to handle only podman not docker but maybe @chavafg can help

**Comment by jodh-intel** [comment]
forward port PR: https://github.com/kata-containers/kata-containers/pull/861

**Pull request merged**